File: app/src/main/java/com/fastaccess/provider/timeline/handler/PreTagHandler.java
package com.fastaccess.provider.timeline.handler;

import android.graphics.Color;
import android.support.annotation.ColorInt;
import android.text.SpannableStringBuilder;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.TypefaceSpan;

import com.fastaccess.helper.PrefGetter;

import net.nightwhistler.htmlspanner.handlers.PreHandler;

import org.htmlcleaner.ContentNode;
import org.htmlcleaner.TagNode;

import lombok.AllArgsConstructor;

import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;

/**
 * Created by Kosh on 22 Apr 2017, 1:07 PM
 */

@AllArgsConstructor public class PreTagHandler extends PreHandler {

    @ColorInt private final int color;
    private final boolean isPre;
    @PrefGetter.ThemeType private int theme;

    private void getPlainText(StringBuffer buffer, Object node) {
        if (node instanceof ContentNode) {
            ContentNode contentNode = (ContentNode) node;
            String text = contentNode.getContent().toString();
            buffer.append(text);
        } else if (node instanceof TagNode) {
            TagNode tagNode = (TagNode) node;
            for (Object child : tagNode.getChildren()) {
                this.getPlainText(buffer, child);
            }
        }
    }

    private String replace(String text) {
        return text.replaceAll("&nbsp;", "\u00A0")
                .replaceAll("&amp;", "&")
                .replaceAll("&quot;", "\"")
                .replaceAll("&cent;", "¢")
                .replaceAll("&lt;", "<")
                .replaceAll("&gt;", ">")
                .replaceAll("&sect;", "§")
                .replaceAll("&ldquo;", "“")
                .replaceAll("&rdquo;", "”")
                .replaceAll("&lsquo;", "‘")
                .replaceAll("&rsquo;", "’")
                .replaceAll("&ndash;", "\u2013")
                .replaceAll("&mdash;", "\u2014")
                .replaceAll("&horbar;", "\u2015");
    }

    @Override public void handleTagNode(TagNode node, SpannableStringBuilder builder, int start, int end) {
        if (isPre) {
            StringBuffer buffer = new StringBuffer();
            buffer.append("\n");//fake padding top + make sure, pre is always by itself
            getPlainText(buffer, node);
            buffer.append("\n");//fake padding bottom + make sure, pre is always by itself
            builder.append(replace(buffer.toString()));
            builder.append("\n");
            builder.setSpan(new CodeBackgroundRoundedSpan(color), start, builder.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.append("\n");
            this.appendNewLine(builder);
            this.appendNewLine(builder);
        } else {
            StringBuffer text = node.getText();
            builder.append(" ");
            builder.append(replace(text.toString()));
            builder.append(" ");
            final int stringStart = start + 1;
            final int stringEnd = builder.length() - 1;
            builder.setSpan(new BackgroundColorSpan(color), stringStart, stringEnd, SPAN_EXCLUSIVE_EXCLUSIVE);
            if (theme == PrefGetter.LIGHT) {
                builder.setSpan(new ForegroundColorSpan(Color.RED), stringStart, stringEnd, SPAN_EXCLUSIVE_EXCLUSIVE);
            }
            builder.setSpan(new TypefaceSpan("monospace"), stringStart, stringEnd, SPAN_EXCLUSIVE_EXCLUSIVE);
        }
    }
}

File: markdown/src/main/java/com/fastaccess/markdown/spans/PreTagHandler.kt
package com.fastaccess.markdown.spans

import android.graphics.Color
import android.text.SpannableStringBuilder
import android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.text.style.TypefaceSpan
import com.fastaccess.github.extensions.isTrue
import net.nightwhistler.htmlspanner.SpanStack
import net.nightwhistler.htmlspanner.TextUtil
import net.nightwhistler.htmlspanner.handlers.PreHandler
import org.htmlcleaner.ContentNode
import org.htmlcleaner.TagNode

/**
 * Created by Kosh on 22 Apr 2017, 1:07 PM
 */
class PreTagHandler(
    private val color: Int = 0,
    private val isPre: Boolean = false,
    private val isLightTheme: Boolean = true
) : PreHandler() {

    private fun getPlainText(buffer: StringBuffer, node: Any?) {
        if (node is ContentNode) {
            val text = TextUtil.replaceHtmlEntities(node.content.toString(), true)
            buffer.append(text)
        } else if (node is TagNode) {
            for (child in node.allChildren) {
                getPlainText(buffer, child)
            }
        }
    }

    private fun replace(text: String): String {
        return text.replace("&nbsp;".toRegex(), "\u00A0")
            .replace("&amp;", "&", true)
            .replace("&quot;", "\"", true)
            .replace("&cent;", "¢", true)
            .replace("&lt;", "<", true)
            .replace("&gt;", ">", true)
            .replace("&sect;", "§", true)
            .replace("&ldquo;", "“", true)
            .replace("&rdquo;", "”", true)
            .replace("&lsquo;", "‘", true)
            .replace("&rsquo;", "’", true)
            .replace("&ndash;", "\u2013", true)
            .replace("&mdash;", "\u2014", true)
            .replace("&horbar;", "\u2015", true)
    }

    override fun beforeChildren(node: TagNode?, builder: SpannableStringBuilder?, spanStack: SpanStack?) {
        super.beforeChildren(node, builder, spanStack)
        isPre.isTrue { node?.addChild(ContentNode("\n")) } // append fake hr
    }

    override fun handleTagNode(node: TagNode?, builder: SpannableStringBuilder?, start: Int, end: Int, spanStack: SpanStack?) {
        builder?.let {
            if (isPre) {
                val buffer = StringBuffer()
                buffer.append("\n")//fake padding top + make sure, pre is always by itself
                getPlainText(buffer, node)
                builder.append(replace(buffer.toString()))
                this.appendNewLine(builder)
                builder.setSpan(CodeBackgroundRoundedSpan(color), start, builder.length, SPAN_EXCLUSIVE_EXCLUSIVE)
            } else {
                val text = node?.text ?: ""
                builder.append(" ")
                builder.append(replace(text.toString()))
                builder.append(" ")
                val stringStart = start + 1
                val stringEnd = builder.length - 1
                builder.setSpan(BackgroundColorSpan(color), stringStart, stringEnd, SPAN_EXCLUSIVE_EXCLUSIVE)
                if (isLightTheme) {
                    builder.setSpan(ForegroundColorSpan(Color.RED), stringStart, stringEnd, SPAN_EXCLUSIVE_EXCLUSIVE)
                }
                builder.setSpan(TypefaceSpan("monospace"), stringStart, stringEnd, SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        }
    }
}
